optimize/guessandcheck.go

```diff
@@ -27,34 +23,63 @@ func (g *GuessAndCheck) Needs() struct{ Gradient, Hessian bool } {
 	return struct{ Gradient, Hessian bool }{false, false}
 }
 
-func (g *GuessAndCheck) Done() {
-	// No cleanup needed
-}
-
 func (g *GuessAndCheck) InitGlobal(dim, tasks int) int {
-	g.eval = make([]bool, tasks)
 	g.bestF = math.Inf(1)
 	g.bestX = resize(g.bestX, dim)
-	g.mux = &sync.Mutex{}
 	return tasks
 }
 
-func (g *GuessAndCheck) IterateGlobal(task int, loc *Location) (Operation, error) {
-	// Task is true if it contains a new function evaluation.
-	if g.eval[task] {
-		g.eval[task] = false
-		g.mux.Lock()
-		if loc.F < g.bestF {
-			g.bestF = loc.F
-			copy(g.bestX, loc.X)
-		} else {
-			loc.F = g.bestF
-			copy(loc.X, g.bestX)
+func (g *GuessAndCheck) sendNewLoc(operation chan<- GlobalTask, task GlobalTask) {
+	g.Rander.Rand(task.X)
+	task.Operation = FuncEvaluation
+	operation <- task
+}
+
+func (g *GuessAndCheck) updateMajor(operation chan<- GlobalTask, task GlobalTask) {
+	// Update the best value seen so far, and send a MajorIteration.
+	if task.F < g.bestF {
+		g.bestF = task.F
+		copy(g.bestX, task.X)
+	} else {
+		task.F = g.bestF
+		copy(task.X, g.bestX)
+	}
+	task.Operation = MajorIteration
+	operation <- task
+}
+
+func (g *GuessAndCheck) RunGlobal(operation chan<- GlobalTask, result <-chan GlobalTask, tasks []GlobalTask) {
+	// Send initial tasks to evaluate
+	for i, task := range tasks {
+		task.Index = i + 1
+		g.sendNewLoc(operation, task)
+	}
+
+	// Read from the channel until PostIteration is sent.
+Outer:
+	for {
+		task := <-result
+		switch task.Operation {
+		default:
+			panic("unknown operation")
+		case PostIteration:
```

Comment:
This is mis/abusing `PostIteration` for a purpose for which it is not intended or documented. I have not read the diff completely yet but this looks something to think about.

Reply:
You're right that the documentation for `PostIteration` needs to be updated, but the intended new behavior is documented https://github.com/gonum/gonum/pull/352/files#diff-a7f524fa342e38670f472a79c2779b82R57